File: apps/mobile/android/app/src/main/java/com/plainva/app/MailNetPlugin.java
package com.plainva.app;

import android.util.Base64;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

/**
 * Raw TCP/TLS sockets for the mail client (mail feinplan G2).
 *
 * Deliberately dumb: open, upgrade to TLS, write bytes, read bytes, close.
 * The IMAP and SMTP protocols live in shared TypeScript, so this file — and
 * its Swift twin — stay small enough to audit, and the two platforms cannot
 * drift apart on protocol behaviour.
 *
 * Certificates are validated by the platform default. There is exactly one
 * exception, mirroring the desktop: a LOOPBACK address may present a
 * self-signed certificate, because that is how the Proton Mail Bridge works
 * and a loopback socket cannot be intercepted.
 */
@CapacitorPlugin(name = "MailNet")
public class MailNetPlugin extends Plugin {

    private static final int MAX_READ = 1 << 16;

    private final Map<String, Conn> conns = new ConcurrentHashMap<>();
    private final AtomicLong ids = new AtomicLong(1);

    private static final class Conn {
        Socket socket;
        InputStream in;
        OutputStream out;
        final String host;
        final int port;

        Conn(Socket socket, String host, int port) throws IOException {
            this.socket = socket;
            this.host = host;
            this.port = port;
            this.in = socket.getInputStream();
            this.out = socket.getOutputStream();
        }
    }

    private static boolean isLoopback(String host) {
        return "localhost".equalsIgnoreCase(host) || "127.0.0.1".equals(host) || "::1".equals(host);
    }

    @PluginMethod
    public void open(PluginCall call) {
        String host = call.getString("host");
        Integer port = call.getInt("port");
        boolean tls = Boolean.TRUE.equals(call.getBoolean("tls", false));
        if (host == null || port == null) {
            call.reject("host and port are required");
            return;
        }
        try {
            Socket raw = new Socket();
            raw.connect(new InetSocketAddress(host, port), 20000);
            raw.setSoTimeout(30000);
            Socket socket = raw;
            if (tls) socket = upgrade(raw, host, port);
            String id = "c" + ids.getAndIncrement();
            conns.put(id, new Conn(socket, host, port));
            JSObject ret = new JSObject();
            ret.put("id", id);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("could not reach the mail server: " + e.getMessage());
        }
    }

    private Socket upgrade(Socket plain, String host, int port) throws IOException {
        SSLSocketFactory factory = isLoopback(host)
                ? LoopbackTls.factory()
                : (SSLSocketFactory) SSLSocketFactory.getDefault();
        SSLSocket tls = (SSLSocket) factory.createSocket(plain, host, port, true);
        tls.setUseClientMode(true);
        if (!isLoopback(host)) {
            // Hostname verification is NOT automatic on a wrapped socket.
            javax.net.ssl.SSLParameters params = tls.getSSLParameters();
            params.setEndpointIdentificationAlgorithm("HTTPS");
            tls.setSSLParameters(params);
        }
        tls.startHandshake();
        tls.setSoTimeout(30000);
        return tls;
    }

    @PluginMethod
    public void startTls(PluginCall call) {
        Conn c = conns.get(call.getString("id", ""));
        if (c == null) {
            call.reject("unknown connection");
            return;
        }
        try {
            Socket tls = upgrade(c.socket, c.host, c.port);
            c.socket = tls;
            c.in = tls.getInputStream();
            c.out = tls.getOutputStream();
            call.resolve();
        } catch (Exception e) {
            call.reject("TLS handshake failed: " + e.getMessage());
        }
    }

    @PluginMethod
    public void write(PluginCall call) {
        Conn c = conns.get(call.getString("id", ""));
        String data = call.getString("data");
        if (c == null || data == null) {
            call.reject("unknown connection");
            return;
        }
        try {
            c.out.write(Base64.decode(data, Base64.DEFAULT));
            c.out.flush();
            call.resolve();
        } catch (IOException e) {
            call.reject("could not send to the mail server: " + e.getMessage());
        }
    }

    @PluginMethod
    public void read(PluginCall call) {
        Conn c = conns.get(call.getString("id", ""));
        if (c == null) {
            call.reject("unknown connection");
            return;
        }
        Integer timeout = call.getInt("timeoutMs", 30000);
        try {
            c.socket.setSoTimeout(timeout == null ? 30000 : timeout);
            byte[] buf = new byte[MAX_READ];
            int n = c.in.read(buf);
            JSObject ret = new JSObject();
            // n < 0 means the peer closed — an empty payload says so upstream.
            ret.put("data", n <= 0 ? "" : Base64.encodeToString(buf, 0, n, Base64.NO_WRAP));
            call.resolve(ret);
        } catch (SocketTimeoutException e) {
            call.reject("the mail server did not answer in time");
        } catch (IOException e) {
            call.reject("lost the connection to the mail server: " + e.getMessage());
        }
    }

    @PluginMethod
    public void close(PluginCall call) {
        Conn c = conns.remove(call.getString("id", ""));
        if (c != null) {
            try {
                c.socket.close();
            } catch (IOException ignored) {
                // Closing a already-dead socket is not an error worth surfacing.
            }
        }
        call.resolve();
    }
}

File: apps/mobile/ios/App/App/MailNetPlugin.swift
import Foundation
import Capacitor
import Network

/**
 * Raw TCP/TLS sockets for the mail client (mail feinplan G2) — the iOS twin of
 * MailNetPlugin.java.
 *
 * Deliberately dumb: open, upgrade to TLS, write bytes, read bytes, close. The
 * IMAP and SMTP protocols live in shared TypeScript, so both platforms run the
 * identical protocol and this file stays small enough to audit.
 *
 * Certificates are validated by the platform default. One exception, mirroring
 * the desktop: a LOOPBACK address may present a self-signed certificate (the
 * Proton Mail Bridge), because a loopback socket cannot be intercepted.
 *
 * A mail server on the user's own network triggers the iOS local-network
 * prompt — `NSLocalNetworkUsageDescription` in Info.plist explains why.
 *
 * `CAPBridgedPlugin` is not optional. Since Capacitor 6 the bridge no longer
 * discovers a plugin's JS name and methods through the Objective-C runtime: it
 * reads them from `jsName` and `pluginMethods`. Without that this class was
 * registered, compiled and shipped — and every call still answered
 * `"MailNet" plugin is not implemented on ios`, because the bridge had no way
 * to connect `registerPlugin("MailNet")` in mailNet.ts to it. Mail was dead on
 * the whole platform (maintainer finding 2026-07-28).
 */
@objc(MailNetPlugin)
public class MailNetPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MailNetPlugin"
    public let jsName = "MailNet"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startTls", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "write", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "read", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise)
    ]

    private final class Conn {
        let connection: NWConnection
        var buffer = Data()
        var closed = false
        init(_ connection: NWConnection) { self.connection = connection }
    }

    private var conns: [String: Conn] = [:]
    private var nextId = 1
    private let lock = NSLock()
    private let queue = DispatchQueue(label: "com.plainva.app.mailnet")

    private static func isLoopback(_ host: String) -> Bool {
        let h = host.lowercased()
        return h == "localhost" || h == "127.0.0.1" || h == "::1"
    }

    private func tlsOptions(host: String) -> NWProtocolTLS.Options {
        let options = NWProtocolTLS.Options()
        if MailNetPlugin.isLoopback(host) {
            // Loopback only — see the class comment.
            sec_protocol_options_set_verify_block(
                options.securityProtocolOptions,
                { _, _, complete in complete(true) },
                queue
            )
        }
        return options
    }

    private func start(_ connection: NWConnection, _ call: CAPPluginCall, id: String) {
        var settled = false
        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                if !settled { settled = true; call.resolve(["id": id]) }
            case .failed(let error):
                if !settled { settled = true; call.reject("could not reach the mail server: \(error.localizedDescription)") }
            case .cancelled:
                if !settled { settled = true; call.reject("the connection was cancelled") }
            default:
                break
            }
        }
        connection.start(queue: queue)
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let host = call.getString("host"), let port = call.getInt("port"), let nwPort = NWEndpoint.Port(rawValue: UInt16(port)) else {
            call.reject("host and port are required")
            return
        }
        let tls = call.getBool("tls") ?? false
        let params: NWParameters = tls
            ? NWParameters(tls: tlsOptions(host: host), tcp: NWProtocolTCP.Options())
            : NWParameters(tls: nil, tcp: NWProtocolTCP.Options())
        let connection = NWConnection(host: NWEndpoint.Host(host), port: nwPort, using: params)

        lock.lock()
        let id = "c\(nextId)"
        nextId += 1
        conns[id] = Conn(connection)
        lock.unlock()

        start(connection, call, id: id)
    }

    /**
     * STARTTLS. Network.framework cannot upgrade a live NWConnection in place,
     * so the connection is replaced by a TLS one to the same endpoint. That is
     * sound for the mail protocols: the server has just answered "OK begin TLS"
     * and expects a fresh handshake, and both IMAP and SMTP re-issue their
     * capability exchange afterwards anyway.
     */
    @objc func startTls(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("unknown connection")
            return
        }
        lock.lock()
        let conn = conns[id]
        lock.unlock()
        guard let conn = conn, case let .hostPort(host, port) = conn.connection.endpoint else {
            call.reject("unknown connection")
            return
        }
        let hostString: String
        switch host {
        case .name(let n, _): hostString = n
        case .ipv4(let a): hostString = "\(a)"
        case .ipv6(let a): hostString = "\(a)"
        @unknown default: hostString = ""
        }
        conn.connection.cancel()
        let params = NWParameters(tls: tlsOptions(host: hostString), tcp: NWProtocolTCP.Options())
        let upgraded = NWConnection(host: host, port: port, using: params)
        let fresh = Conn(upgraded)
        lock.lock()
        conns[id] = fresh
        lock.unlock()
        var settled = false
        upgraded.stateUpdateHandler = { state in
            switch state {
            case .ready:
                if !settled { settled = true; call.resolve() }
            case .failed(let error):
                if !settled { settled = true; call.reject("TLS handshake failed: \(error.localizedDescription)") }
            default:
                break
            }
        }
        upgraded.start(queue: queue)
    }

    @objc func write(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let data = call.getString("data"),
              let bytes = Data(base64Encoded: data) else {
            call.reject("unknown connection")
            return
        }
        lock.lock()
        let conn = conns[id]
        lock.unlock()
        guard let conn = conn else {
            call.reject("unknown connection")
            return
        }
        conn.connection.send(content: bytes, completion: .contentProcessed { error in
            if let error = error {
                call.reject("could not send to the mail server: \(error.localizedDescription)")
            } else {
                call.resolve()
            }
        })
    }

    @objc func read(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("unknown connection")
            return
        }
        lock.lock()
        let conn = conns[id]
        lock.unlock()
        guard let conn = conn else {
            call.reject("unknown connection")
            return
        }
        let timeoutMs = call.getInt("timeoutMs") ?? 30000
        var settled = false
        let finish: (String?, String?) -> Void = { payload, error in
            guard !settled else { return }
            settled = true
            if let error = error { call.reject(error) } else { call.resolve(["data": payload ?? ""]) }
        }
        queue.asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) {
            finish(nil, "the mail server did not answer in time")
        }
        conn.connection.receive(minimumIncompleteLength: 1, maximumLength: 1 << 16) { data, _, isComplete, error in
            if let error = error {
                finish(nil, "lost the connection to the mail server: \(error.localizedDescription)")
                return
            }
            if let data = data, !data.isEmpty {
                finish(data.base64EncodedString(), nil)
                return
            }
            // Empty payload means the peer closed — the shared layer reads that
            // as "connection closed" rather than as an error.
            if isComplete { finish("", nil) }
        }
    }

    @objc func close(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.resolve()
            return
        }
        lock.lock()
        let conn = conns.removeValue(forKey: id)
        lock.unlock()
        conn?.connection.cancel()
        call.resolve()
    }
}
